### PR TITLE
kakoune: support for adding plugins

### PR DESCRIPTION
**Pull request by eraserhd**
###### Motivation for this change

There is a thriving plugin ecosystem for Kakoune now, and it is nice
to add these in our Nix configurations. This was modeled after weechat's
plugins.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

**No binary files to test**
**Closure size: Before: 41,022,312, After: 41,022,312** 

###### Notes

parinfer-rust is useable both standalone and as a Kakoune plugin,
so the plugin file inherits the same definition as pkgs.

I'll make PRs for other plugins if this gets accepted.
[Here](https://github.com/eraserhd/nixpkgs/tree/kak-ansi)'s a tested
branch for the `kak-ansi` plugin.

---

### CODE REVIEW

File: pkgs/applications/editors/kakoune/default.nix
Comment (teto) [inline pkgs/applications/editors/kakoune/default.nix]: you can use `pname = "kakoune";` instead

Comment (eraserhd) [reply pkgs/applications/editors/kakoune/default.nix]: done

**Comment by eraserhd** [comment]
rebased... *bump*

**Comment by teto** [comment]
Does this rebuild kakoune when adding plugins (seems to change buildInputs) ? if that's the case maybe use a wrapper like neovim/neovim-unwrapped, if not ignore my comment.

when that's addressed ping kakune's maintainer.

**Comment by eraserhd** [comment]
OK, I've updated this with the unwrapped/wrapper method.  It's kind of nice, and makes fewer changes.

@vrthra thoughts?

**Comment by eraserhd** [comment]
Hey, no response from @vrthra for 19 days.  I think this is a pretty nice solution now, so will someone else review?

Thanks.

**Comment by Infinisil** [comment]
Ping @alyssais @utdemir

**Review by alyssais — CHANGES_REQUESTED** [review]
Sorry it took me so long to look at this. Meant to ages ago.

Anyway, looks great from here, save from a doc tweak.

File: pkgs/applications/editors/kakoune/plugins.nix
Comment (alyssais) [inline pkgs/applications/editors/kakoune/plugins.nix]:
```suggestion
  inherit parinfer-rust;
```

Comment (eraserhd) [reply pkgs/applications/editors/kakoune/plugins.nix]: Amended the commit and force-pushed for this.  Thank you!

**Comment by eraserhd** [comment]
@alyssais I'm assuming you meant "space tweak" not "doc tweak" (or I'm missing something?), but this should be ready now.

Thanks

**Comment by alyssais** [comment]
> @alyssais I'm assuming you meant "space tweak" not "doc tweak"

Err… yes

**Comment by eraserhd** [comment]
Thank you!

**Comment by jcpetruzza** [comment]
@eraserhd Thanks for adding support for kakoune plugins! I'm trying to use this feature but I'm running into an issue: plugins are sourced *after* one's `kakrc`. This is problematic imo since then one can't require/start/configure those plugins (e.g. the latest `kak-powerline` defines a `powerline-start` command that one needs to run explicitly to activate it).

The source of the problem seems to be the wrapper script, that runs `kak -E "source ../share/kak/plugins.kak" ...`, since the arguments of `-E` will get executed after all rc scripts were executed.

IIUC, the idea was to avoid rebuilding kakoune when the selected plugins change; could this be achieved in some other way? E.g. by using symlinks to the outputs of `kakoune-unwrapped` but changing `$out/share/kak/autoload` to contain both kakoune's rc scripts and the plugins?

**Comment by eraserhd** [comment]
@jcpetruzza Hi!  Yeah, I've noticed this also, after several weeks (or months?) of using it.

The problem is that there isn't a way to tell Kakoune to use a different runtime directory.  If there was, then the wrapper could easily symlink all the original runtime plus plugins into a new runtime, and wrap `kak` to use that runtime.

So, it could be done in two parts, I think:  A PR to kakoune to use `$KAK_RUNTIME`, if it exists, instead of the normal mode of finding the runtime, and then a PR to Nix to use it.   I think that's a pretty nice solution!

Want to try it?  If not, I'll put it on my to-do list, but I'm quite backlogged, so no telling when I will get to it.

**Comment by jcpetruzza** [comment]
I think I got it working in nixpkgs in #91792; the tldr is that we can let nixpkgs make symlinks to everything (kakoune and the plug-ins), and just copy the kak binary, needed for path resolution. The PR is still waiting for review, though...

**Pull request merged**
